src/python/pants/java/nailgun_protocol.py

```diff
@@ -210,14 +215,77 @@ def read_chunk(cls, sock, return_bytes=False):
 
     return chunk_type, payload
 
+  class ProcessStreamTimeout(Exception):
+    """Raised after a timeout completes when a timeout is set on the stream during iteration."""
+
   @classmethod
-  def iter_chunks(cls, sock, return_bytes=False):
-    """Generates chunks from a connected socket until an Exit chunk is sent."""
+  @contextmanager
+  def _set_socket_timeout(cls, sock, timeout=None):
+    """Temporarily set a socket timeout in order to respect a timeout provided to .iter_chunks()."""
+    if timeout is not None:
+      prev_timeout = sock.gettimeout()
+    try:
+      if timeout is not None:
+        sock.settimeout(timeout)
+      yield
+    except socket.timeout:
+      raise cls.ProcessStreamTimeout("socket read timed out with timeout {}".format(timeout))
+    finally:
+      if timeout is not None:
+        sock.settimeout(prev_timeout)
+
+  class TimeoutObject(AbstractClass):
+
+    @abstractmethod
+    def maybe_timeout_options(self):
+      """Called on every stream iteration to obtain a possible specification for a timeout.
+
+      If this method returns non-None, it should return a 2-tuple of the time the timeout began, and
+      the length of the timeout, both as positive floats.
+
+      :returns: a 2-tuple of (timeout_start_time, timeout_interval), or None
+      :rtype: tuple of float
+      """
+
+  @classmethod
+  def iter_chunks(cls, sock, return_bytes=False, timeout_object=None):
+    """Generates chunks from a connected socket until an Exit chunk is sent or a timeout occurs.
+
+    After this method returns non-None once, the .iter_chunks() method will then check the time on
+    every iteration, raising `cls.ProcessStreamTimeout` if the time runs out before the iteration is
+    complete.
+
+    :param sock: the socket to read from.
+    :param bool return_bytes: If False, decode the payload into a utf-8 string.
+    :param cls.TimeoutObject timeout_object: If provided, will be checked every iteration for a
+                                             possible timeout.
+    :raises: :class:`cls.ProcessStreamTimeout`
+    """
+    assert(timeout_object is None or isinstance(timeout_object, cls.TimeoutObject))
+    orig_timeout_time = None
+    timeout_interval = None
     while 1:
-      chunk_type, payload = cls.read_chunk(sock, return_bytes)
-      yield chunk_type, payload
-      if chunk_type == ChunkType.EXIT:
-        break
+      if orig_timeout_time is not None:
+        remaining_time = time.time() - (orig_timeout_time + timeout_interval)
+        if remaining_time < 0:
+          raise cls.ProcessStreamTimeout(
+            "iterating over bytes timed out with timeout interval {} starting at {}, "
```

Comment:
```suggestion
            "iterating over bytes from nailgun timed out with timeout interval {} starting at {}, "
```

(and ideally mention the PID of even better also the name of the process, if possible)

Reply:
I've made this edit, and also made sure to wrap this exception and add the PID in the `_process_session()` method! Working on the process name now.

Reply:
Done! But made into a `logger.warning()` to add the PID and process name while re-raising the original exception.